# DSI Summer Workshops Series

## June 28, 2018

Peggy Lindner<br>
Center for Advanced Computing & Data Science (CACDS)<br>
Data Science Institute (DSI)<br>
University of Houston  
plindner@uh.edu 


This jupyter notebook is available at:
http://130.211.184.150/hub/login


## How Much Money Should Machines Earn? *
### - A journey into computerization (jobs that will be taken over by machines)

Let's learn some R by creating an interactive visualization of some open data because you will train many important skills of a data scientist: 
* loading,  
* transforming and 
* combinig data, 
* cleaning and
* performing a suitable visualization. 

### Datasets used

1. The probability of computerisation of 702 detailed occupations, obtained by Carl Benedikt Frey and Michael A. Osborne from the University of Oxford, using a Gaussian process classifier and published in [this paper](https://www.oxfordmartin.ox.ac.uk/downloads/academic/The_Future_of_Employment.pdf) in 2013.

2. Statistics of jobs from (employments, median annual wages and typical education needed for entry) from the US Bureau of Labor, available here.


In [ ]:
R needs some additional packages to do the work ...

In [2]:
# Load libraries
library(dplyr)
library(tabulizer)
library(rlist)
library(readxl)


Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union



#### Data (Down)Loading

In [3]:
#############################################################################################
# Download and parse data about probability of computerisation
#############################################################################################

# set some variables to be used for download
urlfile <- "https://www.oxfordmartin.ox.ac.uk/downloads/academic/The_Future_of_Employment.pdf"
file <- "The_Future_of_Employment.pdf"

# download the pdf file (if we haven't done so already)
if (!file.exists(file)) {
    download.file(urlfile, destfile = file, mode = 'wb')
}

#### Extracting data from a pdf file

using [Tabula](https://tabula.technology/) from within R

In [4]:
# Extract tables using tabulizer - that looks a little bit like magic( and it takes some time)
out <- extract_tables(file, encoding="UTF-8")

In [5]:
# let's have a look a the "thing" that we just got
out

[[1]]
      [,1] [,2]              [,3] [,4]  [,5]                    [,6] [,7]  [,8]
 [1,] "0"  ""                ""   "0"   ""                      ""   "0"   ""  
 [2,] "0"  "0.5"             "1"  "0"   "0.5"                   "1"  "0"   ""  
 [3,] ""   "Probability of"  ""   ""    "Probability of"        ""   ""    ""  
 [4,] ""   "Computerisation" ""   ""    "Computerisation"       ""   ""    ""  
 [5,] "00" ""                ""   "100" ""                      ""   "80"  ""  
 [6,] ""   ""                ""   ""    "Cramped work space 60" ""   ""    ""  
 [7,] "50" ""                ""   "50"  ""                      ""   "40"  ""  
 [8,] ""   ""                ""   ""    ""                      ""   "20"  ""  
 [9,] "0"  ""                ""   "0"   ""                      ""   "0"   ""  
[10,] "0"  "0.5"             "1"  "0"   "0.5"                   "1"  "0"   ""  
[11,] ""   "Probability of"  ""   ""    "Probability of"        ""   ""    ""  
[12,] ""   "Computerisation" ""   ""    "Computerisation"       ""   ""    ""  
[13,] "80" ""                ""   "80"  ""                      ""   "100" ""  
[14,] "60" ""                ""   "60"  ""                      ""   ""    ""  
[15,] "40" ""                ""   "40"  ""                      ""   "50"  ""  
[16,] "20" ""                ""   "20"  ""                      ""   ""    ""  
[17,] "0"  ""                ""   "0"   ""                      ""   "0"   ""  
[18,] "0"  "0.5"             "1"  "0"   "0.5"                   "1"  "0"   ""  
[19,] ""   "Probability of"  ""   ""    "Probability of"        ""   ""    ""  
[20,] ""   "Computerisation" ""   ""    "Computerisation"       ""   ""    ""  
      [,9]             
 [1,] ""               
 [2,] "0.5 1"          
 [3,] "Probability of" 
 [4,] "Computerisation"
 [5,] ""               
 [6,] "Originalit"     
 [7,] ""               
 [8,] ""               
 [9,] ""               
[10,] "0.5 1"          
[11,] "Probability of" 
[12,] "Computerisation"
[13,] ""               
[14,] ""               
[15,] ""               
[16,] ""               
[17,] ""               
[18,] "0.5 1"          
[19,] "Probability of" 
[20,] "Computerisation"

[[2]]
      [,1]                              [,2]                            
 [1,] "Variable"                        "Probability of Computerisation"
 [2,] ""                                "Low Medium High"               
 [3,] "Assisting and caring for others" "48±20 41±17 34±10"             
 [4,] "Persuasion"                      "48±7.1 35±9.8 32±7.8"          
 [5,] "Negotiation"                     "44±7.6 33±9.3 30±8.9"          
 [6,] "Social perceptiveness"           "51±7.9 41±7.4 37±5.5"          
 [7,] "Fine arts"                       "12±20 3.5±12 1.3±5.5"          
 [8,] "Originality"                     "51±6.5 35±12 32±5.6"           
 [9,] "Manual dexterity"                "22±18 34±15 36±14"             
[10,] "Finger dexterity"                "36±10 39±10 40±10"             
[11,] "Cramped work space"              "19±15 37±26 31±20"             

[[3]]
      [,1]   [,2]             [,3]    [,4]      
 [1,] ""     "Computerisable" ""      ""        
 [2,] "Rank" "Probability"    "Label" "SOC code"
 [3,] "1."   "0.0028"         ""      "29-1125" 
 [4,] "2."   "0.003"          ""      "49-1011" 
 [5,] "3."   "0.003"          ""      "11-9161" 
 [6,] "4."   "0.0031"         ""      "21-1023" 
 [7,] "5."   "0.0033"         ""      "29-1181" 
 [8,] "6."   "0.0035"         ""      "29-1122" 
 [9,] "7."   "0.0035"         ""      "29-2091" 
[10,] "8."   "0.0035"         ""      "21-1022" 
[11,] "9."   "0.0036"         ""      "29-1022" 
[12,] "10."  "0.0036"         ""      "33-1021" 
[13,] "11."  "0.0039"         ""      "29-1031" 
[14,] "12."  "0.0039"         ""      "11-9081" 
[15,] "13."  "0.004"          ""      "27-2032" 
[16,] "14."  "0.0041"         ""      "41-9031" 
[17,] "15."  "0.0042"         "0"     "29-1060" 
[18,] "16."  "0.0042"         ""      

#### Data Transformation

In [7]:
# We are not interested in first two tables - so let's remove them
list.remove(out, c(1:2)) -> tables

# now let's look what we got
tables

[[1]]
      [,1]   [,2]             [,3]    [,4]      
 [1,] ""     "Computerisable" ""      ""        
 [2,] "Rank" "Probability"    "Label" "SOC code"
 [3,] "1."   "0.0028"         ""      "29-1125" 
 [4,] "2."   "0.003"          ""      "49-1011" 
 [5,] "3."   "0.003"          ""      "11-9161" 
 [6,] "4."   "0.0031"         ""      "21-1023" 
 [7,] "5."   "0.0033"         ""      "29-1181" 
 [8,] "6."   "0.0035"         ""      "29-1122" 
 [9,] "7."   "0.0035"         ""      "29-2091" 
[10,] "8."   "0.0035"         ""      "21-1022" 
[11,] "9."   "0.0036"         ""      "29-1022" 
[12,] "10."  "0.0036"         ""      "33-1021" 
[13,] "11."  "0.0039"         ""      "29-1031" 
[14,] "12."  "0.0039"         ""      "11-9081" 
[15,] "13."  "0.004"          ""      "27-2032" 
[16,] "14."  "0.0041"         ""      "41-9031" 
[17,] "15."  "0.0042"         "0"     "29-1060" 
[18,] "16."  "0.0042"         ""      "25-9031" 
[19,] "17."  "0.0043"         ""      "19-3039" 
[20,] "18."  "0.0044"         ""      "33-1012" 
[21,] "19."  "0.0044"         "0"     "29-1021" 
[22,] "20."  "0.0044"         ""      "25-2021" 
[23,] "21."  "0.0045"         ""      "19-1042" 
[24,] "22."  "0.0046"         ""      "11-9032" 
[25,] "23."  "0.0046"         ""      "29-1081" 
[26,] "24."  "0.0047"         ""      "19-3031" 
[27,] "25."  "0.0048"         ""      "21-1014" 
[28,] "26."  "0.0049"         ""      "51-6092" 
[29,] "27."  "0.0055"         ""      "27-1027" 
[30,] "28."  "0.0055"         ""      "11-3121" 
[31,] "29."  "0.0061"         ""      "39-9032" 
[32,] "30."  "0.0063"         ""      "11-3131" 
[33,] "31."  "0.0064"         ""      "29-1127" 
[34,] "32."  "0.0065"         ""      "15-1121" 
[35,] "33."  "0.0067"         "0"     "11-9151" 
[36,] "34."  "0.0068"         ""      "25-4012" 
[37,] "35."  "0.0071"         ""      "29-9091" 
[38,] "36."  "0.0073"         ""      "11-9111" 
[39,] "37."  "0.0074"         "0"     "25-2011" 
[40,] "38."  "0.0075"         ""      "25-9021" 
[41,] "39."  "0.0077"         ""      "19-3091" 
      [,5]                                                            
 [1,] ""                                                              
 [2,] "Occupation"                                                    
 [3,] "Recreational Therapists"                                       
 [4,] "First-Line Supervisors of Mechanics, Installers, and Repairers"
 [5,] "Emergency Management Directors"                                
 [6,] "Mental Health and Substance Abuse Social Workers"              
 [7,] "Audiologists"                                                  
 [8,] "Occupational Therapists"                                       
 [9,] "Orthotists and Prosthetists"                                   
[10,] "Healthcare Social Workers"                                     
[11,] "Oral and Maxillofacial Surgeons"                               
[12,] "First-Line Supervisors of Fire Fighting and Prevention Workers"
[13,] "Dietitians and Nutritionists"                                  
[14,] "Lodging Managers"                                              
[15,] "Choreographers"                                                
[16,] "Sales Engineers"                                               
[17,] "Physicians and Surgeons"                                       
[18,] "Instructional Coordinators"                                    
[19,] "Psychologists, All Other"                                      
[20,] "First-Line Supervisors of Police and Detectives"               
[21,] "Dentists, General"                                             
[22,] "Elementary School Teachers, Except Special Education"          
[23,] "Medical Scientists, Except Epidemiologists"                    
[24,] "Education Administrators, Elementary and Secondary School"     
[25,] "Podiatrists"                                                   
[26,] "Clinical, Counseling, and School Psychologists"                
[27,] "Mental Healt

##### Parse table into something that can be used in the next step

In [8]:
# First we create a placefolder 
prob_comput_df=data.frame()

# Now we go over each of the tables
for (i in 1:length(tables))
{
  # We keep just SOC Code, rank and probability of computerisation
  # We also remove first to lines of each element of table since they are non interesting
  tables[[i]][-c(1,2),c(1,2,4)] %>% 
    as.data.frame(stringsAsFactors = FALSE) %>% 
    rbind(prob_comput_df) -> prob_comput_df
}

In [9]:
# Let's check what we got
prob_comput_df

V1   V2     V3     
1   687. 0.98   43-4151
2   688. 0.98   43-4011
3   689. 0.98   43-9041
4   690. 0.98   51-2093
5   691. 0.99   43-9021
6   692. 0.99   25-4031
7   693. 0.99   43-4141
8   694. 0.99   51-9151
9   695. 0.99   13-2082
10  696. 0.99   43-5011
11  697. 0.99   49-9064
12  698. 0.99   13-2053
13  699. 0.99   15-2091
14  700. 0.99   51-6051
15  701. 0.99   23-2093
16  702. 0.99   41-9041
17  640. 0.96   49-9093
18  641. 0.96   35-2014
19  642. 0.96   39-3031
20  643. 0.96   43-3021
21  644. 0.97   53-6011
22  645. 0.97   51-7042
23  646. 0.97   51-2092
24  647. 0.97   51-6042
25  648. 0.97   51-2023
26  649. 0.97   13-1074
27  650. 0.97   51-6061
28  651. 0.97   51-9081
29  652. 0.97   51-9021
30                     
⋮   ⋮    ⋮      ⋮      
711 10.  0.0036 33-1021
712 11.  0.0039 29-1031
713 12.  0.0039 11-9081
714 13.  0.004  27-2032
715 14.  0.0041 41-9031
716 15.  0.0042 29-1060
717 16.  0.0042 25-9031
718 17.  0.0043 19-3039
719 18.  0.0044 33-1012
720 19.  0.0044 29-1021
721 20.  0.0044 25-2021
722 21.  0.0045 19-1042
723 22.  0.0046 11-9032
724 23.  0.0046 29-1081
725 24.  0.0047 19-3031
726 25.  0.0048 21-1014
727 26.  0.0049 51-6092
728 27.  0.0055 27-1027
729 28.  0.0055 11-3121
730 29.  0.0061 39-9032
731 30.  0.0063 11-3131
732 31.  0.0064 29-1127
733 32.  0.0065 15-1121
734 33.  0.0067 11-9151
735 34.  0.0068 25-4012
736 35.  0.0071 29-9091
737 36.  0.0073 11-9111
738 37.  0.0074 25-2011
739 38.  0.0075 25-9021
740 39.  0.0077 19-3091

In [11]:
# Let's give this thing some proper column names
colnames(prob_comput_df) = c("rank", "probability", "soc")

prob_comput_df

rank probability soc    
1   687. 0.98        43-4151
2   688. 0.98        43-4011
3   689. 0.98        43-9041
4   690. 0.98        51-2093
5   691. 0.99        43-9021
6   692. 0.99        25-4031
7   693. 0.99        43-4141
8   694. 0.99        51-9151
9   695. 0.99        13-2082
10  696. 0.99        43-5011
11  697. 0.99        49-9064
12  698. 0.99        13-2053
13  699. 0.99        15-2091
14  700. 0.99        51-6051
15  701. 0.99        23-2093
16  702. 0.99        41-9041
17  640. 0.96        49-9093
18  641. 0.96        35-2014
19  642. 0.96        39-3031
20  643. 0.96        43-3021
21  644. 0.97        53-6011
22  645. 0.97        51-7042
23  646. 0.97        51-2092
24  647. 0.97        51-6042
25  648. 0.97        51-2023
26  649. 0.97        13-1074
27  650. 0.97        51-6061
28  651. 0.97        51-9081
29  652. 0.97        51-9021
30                          
⋮   ⋮    ⋮           ⋮      
711 10.  0.0036      33-1021
712 11.  0.0039      29-1031
713 12.  0.0039      11-9081
714 13.  0.004       27-2032
715 14.  0.0041      41-9031
716 15.  0.0042      29-1060
717 16.  0.0042      25-9031
718 17.  0.0043      19-3039
719 18.  0.0044      33-1012
720 19.  0.0044      29-1021
721 20.  0.0044      25-2021
722 21.  0.0045      19-1042
723 22.  0.0046      11-9032
724 23.  0.0046      29-1081
725 24.  0.0047      19-3031
726 25.  0.0048      21-1014
727 26.  0.0049      51-6092
728 27.  0.0055      27-1027
729 28.  0.0055      11-3121
730 29.  0.0061      39-9032
731 30.  0.0063      11-3131
732 31.  0.0064      29-1127
733 32.  0.0065      15-1121
734 33.  0.0067      11-9151
735 34.  0.0068      25-4012
736 35.  0.0071      29-9091
737 36.  0.0073      11-9111
738 37.  0.0074      25-2011
739 38.  0.0075      25-9021
740 39.  0.0077      19-3091

In [ ]:
#### Data Cleaning

In [12]:
# what does R think it is looking at?
str(prob_comput_df)

'data.frame':	740 obs. of  3 variables:
 $ rank       : chr  "687." "688." "689." "690." ...
 $ probability: chr  "0.98" "0.98" "0.98" "0.98" ...
 $ soc        : chr  "43-4151" "43-4011" "43-9041" "51-2093" ...


In [13]:
prob_comput_df %>% 
  # convert things that look like numbers into numbers
  mutate(rank=gsub("\\.","", rank) %>% as.numeric()) %>% 
  #let's get rid of missing data
  na.omit() -> prob_comput_df

In [14]:
str(prob_comput_df)

'data.frame':	702 obs. of  3 variables:
 $ rank       : num  687 688 689 690 691 692 693 694 695 696 ...
 $ probability: chr  "0.98" "0.98" "0.98" "0.98" ...
 $ soc        : chr  "43-4151" "43-4011" "43-9041" "51-2093" ...
 - attr(*, "na.action")= 'omit' Named int  30 57 77 92 97 108 112 117 120 125 ...
  ..- attr(*, "names")= chr  "30" "57" "77" "92" ...


In [ ]:
# finally let's delete the file that we just downloaded
file.remove(file)

#### Data (Down)Loading 

In [17]:
#############################################################################################
# Download job statistics
#############################################################################################

# set some variables to be used for download
urlfile <- "https://www.bls.gov/emp/ind-occ-matrix/occupation.xlsx"
file <- "occupation.xlsx"
# Download xlsx file 
if (!file.exists(file)) {
    download.file(urlfile, destfile = file, mode = 'wb')
}

In [18]:
# read excel file into R
job_stats_df <- read_excel(file, 
                           sheet="Table 1.7", 
                           skip=3,
                           col_names = c("job_title",
                                         "soc",
                                         "occupation_type",
                                         "employment_2016",
                                         "employment_2026",
                                         "employment_change_2016_26_nu",
                                         "employment_change_2016_26_pe",
                                         "self_employed_2016_pe",
                                         "occupational_openings_2016_26_av",
                                         "median_annual_wage_2017",
                                         "typical_education_entry",
                                         "work_experience_related_occ",	
                                         "typical_training_needed"))

In [19]:
# now we can remove the downloaded file
file.remove(file)

[1] TRUE

In [22]:
# let's look what we got here
job_stats_df

job_title                                                                                                                                                                                                                                                 
1    Total, all occupations                                                                                                                                                                                                                                    
2    Management occupations                                                                                                                                                                                                                                    
3    Top executives                                                                                                                                                                                                                                            
4    Chief executives                                                                                                                                                                                                                                          
5    General and operations managers                                                                                                                                                                                                                           
6    Legislators                                                                                                                                                                                                                                               
7    Advertising, marketing, promotions, public relations, and sales managers                                                                                                                                                                                  
8    Advertising and promotions managers                                                                                                                                                                                                                       
9    Marketing and sales managers                                                                                                                                                                                                                              
10   Marketing managers                                                                                                                                                                                                                                        
11   Sales managers                                                                                                                                                                                                                                            
12   Public relations and fundraising managers                                                                                                                                                                                                                 
13   Operations specialties managers                                                                                                                                                                                                                           
14   Administrative services managers                                                                                                                                                                                                                          
15   Computer and information systems managers                                                                                                                       

#### Data Transformation & Cleaning

We are going to merge (join) the 2 data sets and keep only the columns that we need.

In [23]:
#############################################################################################
# Join data frames
#############################################################################################
results = prob_comput_df %>% 
  inner_join(job_stats_df, by = "soc") %>% 
  select(job_title, 
         probability, 
         employment_2016, 
         median_annual_wage_2017, 
         typical_education_entry) %>% 
  mutate(probability=as.numeric(probability),
         median_annual_wage_2017=as.numeric(median_annual_wage_2017),
         typical_education_entry=iconv(typical_education_entry, "latin1", "ASCII")) %>% 
  # get rid of missing data
  na.omit()

Warning message in evalq(as.numeric(median_annual_wage_2017), <environment>):
“NAs introduced by coercion”

In [25]:
# Aehmm, can we do that a little slower?
#first, we join using the soc column
first_step <- prob_comput_df %>% 
     inner_join(job_stats_df, by = "soc")

first_step

rank probability soc    
1   687  0.98        43-4151
2   688  0.98        43-4011
3   689  0.98        43-9041
4   690  0.98        51-2093
5   691  0.99        43-9021
6   692  0.99        25-4031
7   693  0.99        43-4141
8   694  0.99        51-9151
9   695  0.99        13-2082
10  696  0.99        43-5011
11  697  0.99        49-9064
12  698  0.99        13-2053
13  699  0.99        15-2091
14  700  0.99        51-6051
15  701  0.99        23-2093
16  702  0.99        41-9041
17  640  0.96        49-9093
18  641  0.96        35-2014
19  642  0.96        39-3031
20  643  0.96        43-3021
21  644  0.97        53-6011
22  645  0.97        51-7042
23  646  0.97        51-2092
24  647  0.97        51-6042
25  648  0.97        51-2023
26  649  0.97        13-1074
27  650  0.97        51-6061
28  651  0.97        51-9081
29  652  0.97        51-9021
30  653  0.97        51-9022
⋮   ⋮    ⋮           ⋮      
658 10   0.0036      33-1021
659 11   0.0039      29-1031
660 12   0.0039      11-9081
661 13   0.004       27-2032
662 14   0.0041      41-9031
663 15   0.0042      29-1060
664 16   0.0042      25-9031
665 17   0.0043      19-3039
666 18   0.0044      33-1012
667 19   0.0044      29-1021
668 20   0.0044      25-2021
669 21   0.0045      19-1042
670 22   0.0046      11-9032
671 23   0.0046      29-1081
672 24   0.0047      19-3031
673 25   0.0048      21-1014
674 26   0.0049      51-6092
675 27   0.0055      27-1027
676 28   0.0055      11-3121
677 29   0.0061      39-9032
678 30   0.0063      11-3131
679 31   0.0064      29-1127
680 32   0.0065      15-1121
681 33   0.0067      11-9151
682 34   0.0068      25-4012
683 35   0.0071      29-9091
684 36   0.0073      11-9111
685 37   0.0074      25-2011
686 38   0.0075      25-9021
687 39   0.0077      19-3091
    job_title                                                                
1   Order clerks                                                             
2   Brokerage clerks                                                         
3   Insurance claims and policy processing clerks                            
4   Timing device assemblers and adjusters                                   
5   Data entry keyers                                                        
6   Library technicians                                                      
7   New accounts clerks                                                      
8   Photographic process workers and processing machine operators            
9   Tax preparers                                                            
10  Cargo and freight agents                                                 
11  Watch repairers                                                          
12  Insurance underwriters                                                   
13  Mathematical technicians                                                 
14  Sewers, hand                                                             
15  Title examiners, abstractors, and searchers                              
16  Telemarketers                                                            
17  Fabric menders, except garment                                           
18  Cooks, restaurant                                                        
19  Ushers, lobby attendants, and ticket takers                              
20  Billing and posting clerks                                               
21  Bridge and lock tenders                                                  
22  Woodworking machine setters, operators, and tenders, except sawing       
23  Team assemblers                                                          
24  Shoe machine operators and tenders                                       
25  Electromechanical equipment assemblers                                   
26  Farm labor contractors                                                   
27  Textile bleaching and dyeing machine operators and tenders               
28  Dental laboratory 

In [26]:
#second, we select only columns that we want
second_step <- first_step %>%
  select(job_title, 
         probability, 
         employment_2016, 
         median_annual_wage_2017, 
         typical_education_entry)

second_step

job_title                                                                
1   Order clerks                                                             
2   Brokerage clerks                                                         
3   Insurance claims and policy processing clerks                            
4   Timing device assemblers and adjusters                                   
5   Data entry keyers                                                        
6   Library technicians                                                      
7   New accounts clerks                                                      
8   Photographic process workers and processing machine operators            
9   Tax preparers                                                            
10  Cargo and freight agents                                                 
11  Watch repairers                                                          
12  Insurance underwriters                                                   
13  Mathematical technicians                                                 
14  Sewers, hand                                                             
15  Title examiners, abstractors, and searchers                              
16  Telemarketers                                                            
17  Fabric menders, except garment                                           
18  Cooks, restaurant                                                        
19  Ushers, lobby attendants, and ticket takers                              
20  Billing and posting clerks                                               
21  Bridge and lock tenders                                                  
22  Woodworking machine setters, operators, and tenders, except sawing       
23  Team assemblers                                                          
24  Shoe machine operators and tenders                                       
25  Electromechanical equipment assemblers                                   
26  Farm labor contractors                                                   
27  Textile bleaching and dyeing machine operators and tenders               
28  Dental laboratory technicians                                            
29  Crushing, grinding, and polishing machine setters, operators, and tenders
30  Grinding and polishing workers, hand                                     
⋮   ⋮                                                                        
658 First-line supervisors of fire fighting and prevention workers           
659 Dietitians and nutritionists                                             
660 Lodging managers                                                         
661 Choreographers                                                           
662 Sales engineers                                                          
663 Physicians and surgeons                                                  
664 Instructional coordinators                                               
665 Psychologists, all other                                                 
666 First-line supervisors of police and detectives                          
667 Dentists, general                                                        
668 Elementary school teachers, except special education                     
669 Medical scientists, except epidemiologists                               
670 Education administrators, elementary and secondary school                
671 Podiatrists                                                              
672 Clinical, counseling, and school psychologists                           
673 Mental health counselors                                                 
674 Fabric and apparel patternmakers                                         
675 Set and exhibit designers                                                
676 Human resources managers                                                 
677 Recreation workers    

In [28]:
#third, we create 2 new columns using the existing columns

third_step <- second_step %>% 
  mutate(probability=as.numeric(probability),
         median_annual_wage_2017=as.numeric(median_annual_wage_2017),
         typical_education_entry=iconv(typical_education_entry, "latin1", "ASCII")) 

third_step

#that looks the same to me, but internally we change some data types
str(second_step)
str(third_step)

Warning message in evalq(as.numeric(median_annual_wage_2017), <environment>):
“NAs introduced by coercion”

job_title                                                                
1   Order clerks                                                             
2   Brokerage clerks                                                         
3   Insurance claims and policy processing clerks                            
4   Timing device assemblers and adjusters                                   
5   Data entry keyers                                                        
6   Library technicians                                                      
7   New accounts clerks                                                      
8   Photographic process workers and processing machine operators            
9   Tax preparers                                                            
10  Cargo and freight agents                                                 
11  Watch repairers                                                          
12  Insurance underwriters                                                   
13  Mathematical technicians                                                 
14  Sewers, hand                                                             
15  Title examiners, abstractors, and searchers                              
16  Telemarketers                                                            
17  Fabric menders, except garment                                           
18  Cooks, restaurant                                                        
19  Ushers, lobby attendants, and ticket takers                              
20  Billing and posting clerks                                               
21  Bridge and lock tenders                                                  
22  Woodworking machine setters, operators, and tenders, except sawing       
23  Team assemblers                                                          
24  Shoe machine operators and tenders                                       
25  Electromechanical equipment assemblers                                   
26  Farm labor contractors                                                   
27  Textile bleaching and dyeing machine operators and tenders               
28  Dental laboratory technicians                                            
29  Crushing, grinding, and polishing machine setters, operators, and tenders
30  Grinding and polishing workers, hand                                     
⋮   ⋮                                                                        
658 First-line supervisors of fire fighting and prevention workers           
659 Dietitians and nutritionists                                             
660 Lodging managers                                                         
661 Choreographers                                                           
662 Sales engineers                                                          
663 Physicians and surgeons                                                  
664 Instructional coordinators                                               
665 Psychologists, all other                                                 
666 First-line supervisors of police and detectives                          
667 Dentists, general                                                        
668 Elementary school teachers, except special education                     
669 Medical scientists, except epidemiologists                               
670 Education administrators, elementary and secondary school                
671 Podiatrists                                                              
672 Clinical, counseling, and school psychologists                           
673 Mental health counselors                                                 
674 Fabric and apparel patternmakers                                         
675 Set and exhibit designers                                                
676 Human resources managers                                                 
677 Recreation workers    

'data.frame':	687 obs. of  5 variables:
 $ job_title              : chr  "Order clerks" "Brokerage clerks" "Insurance claims and policy processing clerks" "Timing device assemblers and adjusters" ...
 $ probability            : chr  "0.98" "0.98" "0.98" "0.98" ...
 $ employment_2016        : num  179 60.4 308.5 0.8 203.8 ...
 $ median_annual_wage_2017: chr  "33510" "49800" "38790" "34800" ...
 $ typical_education_entry: chr  "High school diploma or equivalent" "High school diploma or equivalent" "High school diploma or equivalent" "High school diploma or equivalent" ...
'data.frame':	687 obs. of  5 variables:
 $ job_title              : chr  "Order clerks" "Brokerage clerks" "Insurance claims and policy processing clerks" "Timing device assemblers and adjusters" ...
 $ probability            : num  0.98 0.98 0.98 0.98 0.99 0.99 0.99 0.99 0.99 0.99 ...
 $ employment_2016        : num  179 60.4 308.5 0.8 203.8 ...
 $ median_annual_wage_2017: num  33510 49800 38790 34800 30930 ...
 $ typi

In [31]:
#do we have some missing data points?
is.na(third_step)

job_title probability employment_2016 median_annual_wage_2017
 [1,] FALSE     FALSE       FALSE           FALSE                  
 [2,] FALSE     FALSE       FALSE           FALSE                  
 [3,] FALSE     FALSE       FALSE           FALSE                  
 [4,] FALSE     FALSE       FALSE           FALSE                  
 [5,] FALSE     FALSE       FALSE           FALSE                  
 [6,] FALSE     FALSE       FALSE           FALSE                  
 [7,] FALSE     FALSE       FALSE           FALSE                  
 [8,] FALSE     FALSE       FALSE           FALSE                  
 [9,] FALSE     FALSE       FALSE           FALSE                  
[10,] FALSE     FALSE       FALSE           FALSE                  
[11,] FALSE     FALSE       FALSE           FALSE                  
[12,] FALSE     FALSE       FALSE           FALSE                  
[13,] FALSE     FALSE       FALSE            TRUE                  
[14,] FALSE     FALSE       FALSE           FALSE                  
[15,] FALSE     FALSE       FALSE           FALSE                  
[16,] FALSE     FALSE       FALSE           FALSE                  
[17,] FALSE     FALSE       FALSE           FALSE                  
[18,] FALSE     FALSE       FALSE           FALSE                  
[19,] FALSE     FALSE       FALSE           FALSE                  
[20,] FALSE     FALSE       FALSE           FALSE                  
[21,] FALSE     FALSE       FALSE           FALSE                  
[22,] FALSE     FALSE       FALSE           FALSE                  
[23,] FALSE     FALSE       FALSE            TRUE                  
[24,] FALSE     FALSE       FALSE           FALSE                  
[25,] FALSE     FALSE       FALSE            TRUE                  
[26,] FALSE     FALSE       FALSE           FALSE                  
[27,] FALSE     FALSE       FALSE           FALSE                  
[28,] FALSE     FALSE       FALSE           FALSE                  
[29,] FALSE     FALSE       FALSE           FALSE                  
[30,] FALSE     FALSE       FALSE           FALSE                  
[31,] ⋮         ⋮           ⋮               ⋮                      
[32,] FALSE     FALSE       FALSE           FALSE                  
[33,] FALSE     FALSE       FALSE           FALSE                  
[34,] FALSE     FALSE       FALSE           FALSE                  
[35,] FALSE     FALSE       FALSE           FALSE                  
[36,] FALSE     FALSE       FALSE           FALSE                  
[37,] FALSE     FALSE       FALSE            TRUE                  
[38,] FALSE     FALSE       FALSE           FALSE                  
[39,] FALSE     FALSE       FALSE           FALSE                  
[40,] FALSE     FALSE       FALSE           FALSE                  
[41,] FALSE     FALSE       FALSE           FALSE                  
[42,] FALSE     FALSE       FALSE           FALSE                  
[43,] FALSE     FALSE       FALSE           FALSE                  
[44,] FALSE     FALSE       FALSE           FALSE                  
[45,] FALSE     FALSE       FALSE           FALSE                  
[46,] FALSE     FALSE       FALSE           FALSE                  
[47,] FALSE     FALSE       FALSE            TRUE                  
[48,] FALSE     FALSE       FALSE           FALSE                  
[49,] FALSE     FALSE       FALSE           FALSE                  
[50,] FALSE     FALSE       FALSE           FALSE                  
[51,] FALSE     FALSE       FALSE           FALSE                  
[52,] FALSE     FALSE       FALSE           FALSE                  
[53,] FALSE     FALSE       FALSE           FALSE                  
[54,] FALSE     FALSE       FALSE           FALSE                  
[55,] FALSE     FALSE       FALSE           FALSE                  
[56,] FALSE     FALSE       FALSE           FALSE                  
[57,] FALSE     FALSE       FALSE           FALSE                  
[58,] FALSE     FALSE       FALSE           FALSE             

In [32]:
#show me the rows with missing data
third_step[!complete.cases(third_step),]

job_title                                                                                 
13  Mathematical technicians                                                                  
23  Team assemblers                                                                           
25  Electromechanical equipment assemblers                                                    
89  Electrical and electronic equipment assemblers                                            
174 Medical and clinical laboratory technologists                                             
187 Tour guides and escorts                                                                   
195 Segmental pavers                                                                          
232 Miscellaneous agricultural workers                                                        
237 Buyers and purchasing agents, farm products                                               
246 Purchasing agents, except wholesale, retail, and farm products                            
367 Computer support specialists                                                              
381 First-line supervisors of helpers, laborers, and material movers, hand                    
390 Medical and clinical laboratory technicians                                               
411 Slot supervisors                                                                          
429 Gaming supervisors                                                                        
431 Wholesale and retail buyers, except farm products                                         
452 Actors                                                                                    
515 Travel guides                                                                             
527 Musicians and singers                                                                     
551 Dancers                                                                                   
562 Orthodontists                                                                             
573 First-line supervisors of transportation and material-moving machine and vehicle operators
580 Postsecondary teachers                                                                    
582 Substance abuse and behavioral disorder counselors                                        
657 Oral and maxillofacial surgeons                                                           
663 Physicians and surgeons                                                                   
673 Mental health counselors                                                                  
    probability employment_2016 median_annual_wage_2017
13  0.9900         0.6             NA                  
23  0.9700      1130.9             NA                  
25  0.9700        45.7             NA                  
89  0.9500       218.9             NA                  
174 0.9000       171.4             NA                  
187 0.9100        45.8             NA                  
195 0.8300         2.1             NA                  
232 0.8700       847.3          23710                  
237 0.8700        13.7             NA                  
246 0.7700       309.4             NA                  
367 0.6500       835.3          52810                  
381 0.4200       184.4             NA                  
390 0.4700       164.2             NA                  
411 0.5400        12.1             NA                  
429 0.2800        38.5             NA                  
431 0.2900       123.3             NA                  
452 0.3700        63.8             NA                  
515 0.0570         3.6             NA                  
527 0.0740       172.4             NA                  
551 0.1300        13.5             NA                  
562 0.0230         6.6             NA                  
573 0.0290       204.2             NA                  
580 0.0320      1871.4          67140                  


In [30]:
# and last but not least we remove the rows with missing data
results <- third_step %>%
    na.omit()

job_title probability employment_2016 median_annual_wage_2017
 [1,] FALSE     FALSE       FALSE           FALSE                  
 [2,] FALSE     FALSE       FALSE           FALSE                  
 [3,] FALSE     FALSE       FALSE           FALSE                  
 [4,] FALSE     FALSE       FALSE           FALSE                  
 [5,] FALSE     FALSE       FALSE           FALSE                  
 [6,] FALSE     FALSE       FALSE           FALSE                  
 [7,] FALSE     FALSE       FALSE           FALSE                  
 [8,] FALSE     FALSE       FALSE           FALSE                  
 [9,] FALSE     FALSE       FALSE           FALSE                  
[10,] FALSE     FALSE       FALSE           FALSE                  
[11,] FALSE     FALSE       FALSE           FALSE                  
[12,] FALSE     FALSE       FALSE           FALSE                  
[13,] FALSE     FALSE       FALSE            TRUE                  
[14,] FALSE     FALSE       FALSE           FALSE                  
[15,] FALSE     FALSE       FALSE           FALSE                  
[16,] FALSE     FALSE       FALSE           FALSE                  
[17,] FALSE     FALSE       FALSE           FALSE                  
[18,] FALSE     FALSE       FALSE           FALSE                  
[19,] FALSE     FALSE       FALSE           FALSE                  
[20,] FALSE     FALSE       FALSE           FALSE                  
[21,] FALSE     FALSE       FALSE           FALSE                  
[22,] FALSE     FALSE       FALSE           FALSE                  
[23,] FALSE     FALSE       FALSE            TRUE                  
[24,] FALSE     FALSE       FALSE           FALSE                  
[25,] FALSE     FALSE       FALSE            TRUE                  
[26,] FALSE     FALSE       FALSE           FALSE                  
[27,] FALSE     FALSE       FALSE           FALSE                  
[28,] FALSE     FALSE       FALSE           FALSE                  
[29,] FALSE     FALSE       FALSE           FALSE                  
[30,] FALSE     FALSE       FALSE           FALSE                  
[31,] ⋮         ⋮           ⋮               ⋮                      
[32,] FALSE     FALSE       FALSE           FALSE                  
[33,] FALSE     FALSE       FALSE           FALSE                  
[34,] FALSE     FALSE       FALSE           FALSE                  
[35,] FALSE     FALSE       FALSE           FALSE                  
[36,] FALSE     FALSE       FALSE           FALSE                  
[37,] FALSE     FALSE       FALSE            TRUE                  
[38,] FALSE     FALSE       FALSE           FALSE                  
[39,] FALSE     FALSE       FALSE           FALSE                  
[40,] FALSE     FALSE       FALSE           FALSE                  
[41,] FALSE     FALSE       FALSE           FALSE                  
[42,] FALSE     FALSE       FALSE           FALSE                  
[43,] FALSE     FALSE       FALSE           FALSE                  
[44,] FALSE     FALSE       FALSE           FALSE                  
[45,] FALSE     FALSE       FALSE           FALSE                  
[46,] FALSE     FALSE       FALSE           FALSE                  
[47,] FALSE     FALSE       FALSE            TRUE                  
[48,] FALSE     FALSE       FALSE           FALSE                  
[49,] FALSE     FALSE       FALSE           FALSE                  
[50,] FALSE     FALSE       FALSE           FALSE                  
[51,] FALSE     FALSE       FALSE           FALSE                  
[52,] FALSE     FALSE       FALSE           FALSE                  
[53,] FALSE     FALSE       FALSE           FALSE                  
[54,] FALSE     FALSE       FALSE           FALSE                  
[55,] FALSE     FALSE       FALSE           FALSE                  
[56,] FALSE     FALSE       FALSE           FALSE                  
[57,] FALSE     FALSE       FALSE           FALSE                  
[58,] FALSE     FALSE       FALSE           FALSE             

In [33]:
#what did we get?
results

job_title                                                                
1   Order clerks                                                             
2   Brokerage clerks                                                         
3   Insurance claims and policy processing clerks                            
4   Timing device assemblers and adjusters                                   
5   Data entry keyers                                                        
6   Library technicians                                                      
7   New accounts clerks                                                      
8   Photographic process workers and processing machine operators            
9   Tax preparers                                                            
10  Cargo and freight agents                                                 
11  Watch repairers                                                          
12  Insurance underwriters                                                   
14  Sewers, hand                                                             
15  Title examiners, abstractors, and searchers                              
16  Telemarketers                                                            
17  Fabric menders, except garment                                           
18  Cooks, restaurant                                                        
19  Ushers, lobby attendants, and ticket takers                              
20  Billing and posting clerks                                               
21  Bridge and lock tenders                                                  
22  Woodworking machine setters, operators, and tenders, except sawing       
24  Shoe machine operators and tenders                                       
26  Farm labor contractors                                                   
27  Textile bleaching and dyeing machine operators and tenders               
28  Dental laboratory technicians                                            
29  Crushing, grinding, and polishing machine setters, operators, and tenders
30  Grinding and polishing workers, hand                                     
31  Pesticide handlers, sprayers, and applicators, vegetation                
32  Log graders and scalers                                                  
33  Ophthalmic laboratory technicians                                        
⋮   ⋮                                                                        
655 Orthotists and prosthetists                                              
656 Healthcare social workers                                                
658 First-line supervisors of fire fighting and prevention workers           
659 Dietitians and nutritionists                                             
660 Lodging managers                                                         
661 Choreographers                                                           
662 Sales engineers                                                          
664 Instructional coordinators                                               
665 Psychologists, all other                                                 
666 First-line supervisors of police and detectives                          
667 Dentists, general                                                        
668 Elementary school teachers, except special education                     
669 Medical scientists, except epidemiologists                               
670 Education administrators, elementary and secondary school                
671 Podiatrists                                                              
672 Clinical, counseling, and school psychologists                           
674 Fabric and apparel patternmakers                                         
675 Set and exhibit designers                                                
676 Human resources managers                                                 
677 Recreation workers    

#### Finally, let's create a visualization

We are ging to use [Highcharter](http://jkunst.com/highcharter/index.html) which is just one of many ways to create interactive visualizations in R.

In [34]:
#we need some more packages 
library(highcharter)
library(htmlwidgets)
library(IRdisplay)

Highcharts (www.highcharts.com) is a Highsoft software product which is
not free for commercial and Governmental use


In [37]:
#let's create an object that is actually a visual
x=hchart(results, 
       "scatter", 
       hcaes(x = probability*100, 
             y = median_annual_wage_2017, 
             group=typical_education_entry, 
             size=employment_2016)) %>% 
  hc_title(text = "How Much Money Should Machines Earn?") %>%
  hc_subtitle(text = "Probability of Computerisation and Wages by Job") %>% 
  hc_credits(enabled = TRUE, text = "Source: Oxford Martin School and US Department of Labor") %>% 
  hc_xAxis(title = list(text = "Probability of Computerisation"), labels = list(format = "{value}%")) %>% 
  hc_yAxis(title = list(text = "Median Annual Wage 2017"), labels = list(format = "{value}$")) %>% 
  hc_plotOptions(bubble = list(minSize = 3, maxSize = 35)) %>% 
  hc_tooltip(formatter = JS("function(){
                            return ('<b>'+ this.point.job_title + '</b><br>'+
                            'Probability of computerisation: '+ Highcharts.numberFormat(this.x, 0)+'%' + 
                            '<br>Median annual wage 2017 ($): '+ Highcharts.numberFormat(this.y, 0) + 
                            '<br>Employment 2016 (000s): '+ Highcharts.numberFormat(this.point.size, 0) )}")) %>% 
  hc_chart(zoomType = "xy") %>%
  hc_exporting(enabled = TRUE)

# it's an object!
str(x)

List of 8
 $ x            :List of 6
  ..$ hc_opts  :List of 11
  .. ..$ title             :List of 1
  .. .. ..$ text: chr "How Much Money Should Machines Earn?"
  .. ..$ yAxis             :List of 3
  .. .. ..$ title :List of 1
  .. .. .. ..$ text: chr "Median Annual Wage 2017"
  .. .. ..$ type  : chr "linear"
  .. .. ..$ labels:List of 1
  .. .. .. ..$ format: chr "{value}$"
  .. ..$ credits           :List of 2
  .. .. ..$ enabled: logi TRUE
  .. .. ..$ text   : chr "Source: Oxford Martin School and US Department of Labor"
  .. ..$ exporting         :List of 1
  .. .. ..$ enabled: logi TRUE
  .. ..$ plotOptions       :List of 4
  .. .. ..$ series :List of 3
  .. .. .. ..$ turboThreshold: num 0
  .. .. .. ..$ showInLegend  : logi TRUE
  .. .. .. ..$ marker        :List of 1
  .. .. .. .. ..$ enabled: logi TRUE
  .. .. ..$ treemap:List of 1
  .. .. .. ..$ layoutAlgorithm: chr "squarified"
  .. .. ..$ bubble :List of 2
  .. .. .. ..$ minSize: num 3
  .. .. .. ..$ maxSize: num 35
  .. 

In [38]:
# and now let's get this object showing up in our jupyter notebook
saveWidget(x, 'demox.html', selfcontained = FALSE)
display_html('<iframe src="demox.html", width = 900, height = 500 ></iframe>')

A full size version of the visualization can be found [here](https://fronkonstin.com/wp-content/uploads/2018/06/machines_wage.html)


And thanks again to the person who wrote the [original post](https://fronkonstin.com/2018/06/17/how-much-money-should-machines-earn/)!

These are some insights:

* There is a moderate negative correlation between wages and probability of computerisation.
* Around 45% of US employments are threatened by machines (have a computerisation probability higher than 80%): half of them do not require formal education to entry.
* In fact, 78% of jobs which do not require formal education to entry are threatened by machines: 0% which require a master’s degree are.
* Teachers are absolutely irreplaceable (0% are threatened by machines) but they earn a 2.2% less then the average wage (unfortunately, I’m afraid this phenomenon occurs in many other countries as well).
* Don’t study for librarian or archivist: it seems a bad way to invest your time
* Mathematicians will survive to machines

##### What do you see there?